src/Tool/Commands/ServiceControlCommand.cs

```diff
@@ -122,25 +124,181 @@ async Task<QueueThroughput[]> SampleData(int minutes, CancellationToken cancella
+        {
+            var middlePageNum = (minPage + maxPage) / 2;
+            var pageData = await GetPage(middlePageNum);
+
+            // If we've backtracked to a page we've hit before, or the page actually contains the time we seek, we're done
+            if (middlePageNum == minPage || middlePageNum == maxPage || pageData.ContainsTime(sampleStart))
+            {
+                Console.WriteLine($"  * Found => {(auditSamplingPageSize * (middlePageNum - 1)) + pageData.CountGreaterThan(sampleStart)} messages");
+                return new QueueThroughput
+                {
+                    QueueName = endpointName,
+                    Throughput = (auditSamplingPageSize * (middlePageNum - 1)) + pageData.CountGreaterThan(sampleStart)
+                };
+            }
+
+            if (pageData.Count == 0 || pageData.ProcessedAtMax < sampleStart)
+            {
+                // Went too far, cut out the top half
+                maxPage = middlePageNum;
+            }
+            else if (pageData.ProcessedAtMin > sampleStart)
+            {
+                // Not far enough, cut out the bottom half
+                minPage = middlePageNum;
+            }
+        }
+
+        // Likely we don't get here, but for completeness
+        var finalPage = await GetPage(minPage);
+        Console.WriteLine($"  * Catch-All => {(auditSamplingPageSize * (minPage - 1)) + finalPage.CountGreaterThan(sampleStart)} messages");
+        return new QueueThroughput
+        {
+            QueueName = endpointName,
+            Throughput = (auditSamplingPageSize * (minPage - 1)) + finalPage.CountGreaterThan(sampleStart)
+        };
+    }
+
+    async Task<AuditBatch> GetAuditBatch(string endpointName, int page, int pageSize, CancellationToken cancellationToken)
+    {
+        var url = $"{primaryUrl}/endpoints/{endpointName}/messages/?page={page}&per_page={pageSize}&sort=processed_at&direction=desc";
+
+        var arr = await GetServiceControlData<JArray>(url, cancellationToken);
+
+        var processedAtValues = arr.Select(token => token["processed_at"].Value<DateTime>()).ToArray();
+
+        return new AuditBatch(processedAtValues);
+    }
+
+    record struct AuditBatch
+    {
+        public AuditBatch(DateTime[] timestamps)
+        {
+            this.timestamps = timestamps;
+            Count = timestamps.Length;
+
+            if (Count > 0)
+            {
+                ProcessedAtMin = timestamps.Min();
+                ProcessedAtMax = timestamps.Max();
+                EstimatedSecPerMsg = (ProcessedAtMax - ProcessedAtMin).TotalSeconds / Count;
+            }
+            else
+            {
+                ProcessedAtMin = default;
+                ProcessedAtMax = default;
+                EstimatedSecPerMsg = 0;
+            }
+        }
+
+        DateTime[] timestamps;
+
+        public int Count { get; }
+        public DateTime ProcessedAtMin { get; }
+        public DateTime ProcessedAtMax { get; }
```

Comment:
Suggestion: instead of ProcesssedAtMin and Max would we use FirstMessageProcessedAt and LastMessageProcessedAt? I think that would make this stuff easier to reason about, but feel free to disagree.